lib/dep-graph-builders/pnpm/lockfile-parser/lockfile-parser.ts

```diff
@@ -0,0 +1,162 @@
+          optionalDependencies: versionData.dependencies,
+        };
+        packages[`${pkg.name}@${pkg.version}`] = pkg;
+      },
+    );
+    return packages;
+  }
+
+  public extractTopLevelDependencies(options: {
+    includeDevDeps: boolean;
+    includeOptionalDeps?: boolean;
+    includePeerDeps?: boolean;
+  }): PnpmDeps {
+    const prodDeps = this.normalizeTopLevelDeps(this.dependencies || {}, false);
+    const devDeps = options.includeDevDeps
+      ? this.normalizeTopLevelDeps(this.devDependencies || {}, true)
+      : {};
+
+    const optionalDeps = options.includeOptionalDeps
+      ? this.normalizeTopLevelDeps(this.optionalDependencies || {}, false)
+      : {};
+
+    const peerDeps = options.includePeerDeps
+      ? this.normalizeTopLevelDeps(this.peerDependencies || {}, false)
+      : {};
+
+    return { ...prodDeps, ...devDeps, ...optionalDeps, ...peerDeps };
+  }
+
+  public normalizeVersion(
+    name: string,
+    version: string,
+    isDev: boolean,
+  ): string {
+    if (this.isWorkspaceLockfile()) {
+      version = this.resolveWorkspacesCrossReference(name, version, isDev);
+    }
+    if (!valid(version)) {
+      version = this.excludeTransPeerDepsVersions(version);
+      if (!valid(version)) {
+        // for npm and git ref packages
+        // they show up in packages with keys equal to the version in top level deps
+        // e.g. body-parser with version github.com/expressjs/body-parser/263f602e6ae34add6332c1eb4caa808893b0b711
+        if (this.packages[version]) {
+          return this.packages[version].version!;
+        }
+      }
+    }
+    return version;
+  }
+
+  public resolveWorkspacesCrossReference(
+    name: string,
+    version: string,
+    isDev: boolean,
+  ): string {
+    if (version.startsWith('link:')) {
+      // In  workspaces example:
+      // package-b:
+      //   specifier: 1.0.0
+      //   version: link:../pkg-b
+      const depPath = version.split('link:')[1];
+      const resolvedPathDep = pathUtil.join(
+        this.workspaceArgs?.workspacePath!,
+        depPath,
+      );
+      // cross referenced package, we add it to the extracted packages
+      version = this.workspaceArgs?.projectsVersionMap[resolvedPathDep]!;
+
+      // todo: consider getting devDependencies, optionaldependencies
+      const resolvedDeps = this.normalizePackagesDeps(
+        this.rawPnpmLock.importers[resolvedPathDep].dependencies,
+        isDev,
+      );
+
+      this.extractedPackages[`${name}@${version}`] = {
+        name,
+        version,
+        id: `${name}@${version}`,
+        isDev,
+        dependencies: resolvedDeps,
+      };
+    }
+    return version;
+  }
+
+  abstract normalizePackagesDeps(dependencies, isDev): Record<string, string>;
+
+  abstract normalizeTopLevelDeps(dependencies, isDev): PnpmDeps;
+
+  abstract parseDepPath(depPath: string);
+
+  abstract excludeTransPeerDepsVersions(fullVersionStr: string): string;
+}
```

Comment:
Interesting 😅 

Can you elaborate on the reason for this kind of abstraction here?